src/lib/scheduler/task_queue.cpp

```diff
@@ -33,46 +33,57 @@ void TaskQueue::push(const std::shared_ptr<AbstractTask>& task, const SchedulePr
   }
 
   task->set_node_id(_node_id);
-  _queues[priority_uint].push(task);
-
-  new_task.notify_one();
+  [[maybe_unused]] const auto enqueue_successful = _queues[priority_uint].enqueue(task);
+  DebugAssert(enqueue_successful, "Enqueuing did not succeed.");
+  semaphore.signal();
 }
 
 std::shared_ptr<AbstractTask> TaskQueue::pull() {
-  std::shared_ptr<AbstractTask> task;
+  auto task = std::shared_ptr<AbstractTask>{};
   for (auto& queue : _queues) {
-    if (queue.try_pop(task)) {
+    if (queue.try_dequeue(task)) {
       return task;
     }
   }
+
+  // We waited for the semaphore to enter pull() but did not receive a task. Ensure that queues are checked again.
+  semaphore.signal();
   return nullptr;
 }
 
 std::shared_ptr<AbstractTask> TaskQueue::steal() {
-  std::shared_ptr<AbstractTask> task;
+  auto task = std::shared_ptr<AbstractTask>{};
   for (auto& queue : _queues) {
-    if (queue.try_pop(task)) {
+    if (queue.try_dequeue(task)) {
       if (task->is_stealable()) {
```

Comment:
Are shutdown tassk stealable? They should not be under any circumstances, right? Should also be eased by using a dedicated class.